Rearange whitebit service and connector layers maksym khomishchak

mkhomishchak opened the pull request:
nstead of running mapping and any other logic from integration level itself, I moved it all to the connector service that is a level higher one
now WhiteBitService will be responsible only for getting and returning responses to the connector level, addition logic or error validation for responses should be implemented on connector level
added error handling for the whitebit exceptions, updated unit tests

mkhomishchak committed:
refactore WhiteBit integration service layer

instead of running mapping and any other logic from integration level itself, I moved it all to the connector service that is a level higher one
now WhiteBitService will be responsible only for getting and returning responses to the connector level, addition logic or error validation for responses should be implemented on connector level
need to add error validation and tests for testing error validation

mkhomishchak committed:
add Mono error handling

added error handling for the whitebit exceptions, updated unit tests

mkhomishchak committed:
remove unused bean

mkhomishchak committed:
remove unused classes

mkhomishchak committed:
uncomment integration tests

mkhomishchak committed:
remove unused import

DendeberiaOleksandr commented on src/main/java/com/khomishchak/ws/services/integration/whitebit/WhiteBitExchangerConnectorService.java:
```diff
@@ -1,26 +1,107 @@
+                .orElseGet(() -> getNewExchangerDepositWithdrawalTransactions(balance));
+    }
+
+    private ExchangerDepositWithdrawalTransactions getNewExchangerDepositWithdrawalTransactions(Balance balance) {
```
Isn't better to move it to entity layer ?

mkhomishchak replied:
I prefer to not to have any business related logic on entity layer

DendeberiaOleksandr commented on src/main/java/com/khomishchak/ws/services/integration/whitebit/WhiteBitExchangerConnectorService.java:
```diff
@@ -1,26 +1,107 @@
+        return transactions;
+    }
+
+    private void assigneeTransactionsToExchangerTransactionsEntity(ExchangerDepositWithdrawalTransactions exchangerTransactions,
```
Isn't better to move it to entity layer ?

mkhomishchak replied:
I prefer to not to have any business related logic on entity layer

DendeberiaOleksandr commented on src/main/java/com/khomishchak/ws/services/integration/whitebit/WhiteBitServiceImpl.java:
```diff
@@ -221,6 +157,20 @@ private DecryptedApiKeySettingDTO getApiKeysPair(long userId) {
     }
 
+    private <T> Mono<T> getErrorResp(WhiteBitV4ErrorResp resp, int statusCode) {
+        List<String> errors = resp.getErrors() != null ? resp.getErrors().getMessages() : new ArrayList<>();
```
```suggestion
        Object errors = resp.getErrors();
        List<String> errorMessages = errors != null ? errors.getMessages() : new ArrayList<>();
```
Also it can be done by `Optional` but it's on  your taste

mkhomishchak committed:
refactor creation of error response

mkhomishchak committed:
add java 8 time API support

mkhomishchak committed:
replaced api keys adapter with custom repositrory interface

replaced api keys adapter that directly injects repository with interface that extends both: custom and native method that are needed for api keys

mkhomishchak committed:
move transactions logic to entity layer

mkhomishchak committed:
refactor constructor for ExchangerTransactions

DendeberiaOleksandr commented on src/main/java/com/khomishchak/ws/controllers/AbstractRestController.java:
```diff
@@ -29,4 +30,13 @@ public ResponseEntity<?> handleBalanceNotFoundException(BalanceNotFoundException
+
+    @ExceptionHandler(WhiteBitClientException.class)
+    public ResponseEntity<?> handleBalanceNotFoundException(WhiteBitClientException ex) {
+        Map<String, Object> result = new HashMap<>();
```
You can replace it by some `ErrorResponse` class etc. (of course created by you). As I can see you used a map before with key "errors" so would be great to do it.

mkhomishchak replied:
they have different format

DendeberiaOleksandr replied:
They are same, aren't they ? The only difference is here you include three fields instead of single one. But you can create an error representation and annotate it with `@JsonInclude(NON_NULL)`

DendeberiaOleksandr commented on src/main/java/com/khomishchak/ws/repositories/custom/impl/ApiKeySettingCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,46 @@
+                .build();
+    }
+
+    private String decryptKey(String encryptedKey) {
```
Sounds like it's not a responsibility of repository

mkhomishchak replied:
it is a wrapper class for spring native jpa repository interface 
it will return already decrypted keys and will also provide all range of native repo interface

DendeberiaOleksandr replied:
I mean sounds like Dao should not be responsible for the decryption. Move it to some service.

DendeberiaOleksandr commented on src/main/java/com/khomishchak/ws/services/integration/whitebit/configs/WhiteBitWebConfig.java:
```diff
@@ -34,10 +35,16 @@ HttpClient whiteBitApiHttpClient(
 
     @Qualifier("WhiteBitApiWebClient")
     @Bean
```
Wrong usage of `@Qualifier`. This annotation is only for injection purposes. So if you want to set a bean id manually you should provide it as a parameter of `@Bean` instead.
```suggestion
    @Bean("WhiteBitApiWebClient")
```

DendeberiaOleksandr commented on src/main/java/com/khomishchak/ws/services/integration/whitebit/configs/WhiteBitWebConfig.java:
```diff
@@ -34,10 +35,16 @@ HttpClient whiteBitApiHttpClient(
     }
+
+    @Qualifier("WhiteBitObjectMapper")
+    @Bean
```
```suggestion
    @Bean("WhiteBitObjectMapper")
```

mkhomishchak committed:
removed hibernate detached issue, updated usage of @Qualifier

we did have the issue once persisting ExchangerDepositWithdrawalTransactions, the problem was that Balance inside it was in detached state because of the .block() call
removed usage of @Qualifier for bean creation
updated time format for transactions

mkhomishchak committed:
Use single error resp instance instead of Map

mkhomishchak committed:
removed jpa decorators and instead used separated service related to api key setting entity

separated conserns and instead of calling dao decorator that implement additional decryption logic, and instead created a service related api key settings only, and it will handle the decryption

mkhomishchak committed:
fix unit tests after introducing new service layer instead of dao
